docs/design/eth-subscribe.md

```diff
@@ -0,0 +1,332 @@
+|:-------------:|:------------------------------------------------------------------------------------------------------------------------:|:---------------------------------------------------------------------------------------------------------------------------------------:|
+|     32607     | The maximum batch size that can be sent over a websocket connection is 10. Please decrease the batch size and try again. |                             Occurs when user attempts to send high-volume JSON-RPC traffic over Websockets.                             |
+|     32608     |                       The maximum number of socket connections is reached. Please try again later.                       |     Occurs when a new connection is established, but that exceeds the limit of allowed live connections (controlled by an env var).     |
+|     32609     |                                                 Invalid subscription id                                                  | Occurs when user tries to unsubscribe from a non-existing subscription or one that is not associated with the active socket connection. |
+|     32610     |                                    Invalid subscription parameters: {detailed reason}                                    |               Occurs when there is an error due to the provided subscription parameters, i.e. the address does not exist.               |
+|     32611     |                                                   Subscription aborted                                                   |                          Occurs when the subscription is aborted by the relay, i.e. a maximum TTL is reached.                           |
+
+## Non-Functional Requirements
+
+WebSockets in general are resource intensive. We could implement a throttling mechanism. Every live socket connection would also result in additional requests to the Mirror Node every n seconds. This means that we need to introduce a rate limiting feature, like 5 concurrent connections by IP.
+
+1. Implement an IP based limit for maximum concurrent connections.
+2. Implement a global limit for maximum concurrent connections.
+3. Explore the possibility of adding a TTL limit for every connection.
+4. Explore the option to add a limit for the total subscribed time per IP ( `sum(connection * subscribed_time)` ). There could theoretically be a way to charge users to increase this limit, for example through via a smart contract. Alternatively users could be encouraged to run their own instances of the relay.
+5. Users should be required to renew their subscription in the case of an error. If for some reason the relay is restarted it should not automatically try to restore all previous connections and subscriptions.
+
+## Open Questions
+
+1. How will we limit connections and subscriptions?
+2. Is it reasonable to add a TTL timer to subscriptions?
+
```

Comment:
Have we thought through the expectations around subscriptions and performance?  Ethereum's block times are approximately 13 seconds.  Will this architecture support the polling of the mirror node every 13 seconds with out triggering rate limits and throttling?  I am asking because some of this transaction related event information can be extremely time sensitive when maintaining asset prices on various protocols, e.g. arbitrage between liquidity pools, oracles adjusting prices based on external events, etc.

Reply:
I'm not sure what the exact rate limits are for the mirror node, but I did a quick test and it can easily handle at least 100 requests per second.

Reply:
Okay, that's good news.